presto-hive/src/main/java/com/facebook/presto/hive/FilteringPageSource.java

```diff
@@ -0,0 +1,289 @@
+            TupleDomain<HiveColumnHandle> domainPredicate,
+            RowExpression remainingPredicate,
+            TypeManager typeManager,
+            RowExpressionService rowExpressionService,
+            ConnectorSession session,
+            Set<Integer> originalIndexes,
+            ConnectorPageSource delegate)
+    {
+        requireNonNull(rowExpressionService, "rowExpressionService is null");
+        requireNonNull(remainingPredicate, "remainingPredicate is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        this.delegate = requireNonNull(delegate, "delegate is null");
+
+        domainFilters = new TupleDomainFilter[columnMappings.size()];
+        columnTypes = new Type[columnMappings.size()];
+        if (!domainPredicate.isAll()) {
+            Map<Integer, Domain> domains = domainPredicate.transform(HiveColumnHandle::getHiveColumnIndex).getDomains().get();
+            for (int i = 0; i < columnMappings.size(); i++) {
+                HiveColumnHandle columnHandle = columnMappings.get(i).getHiveColumnHandle();
+                int hiveColumnIndex = columnHandle.getHiveColumnIndex();
+                if (domains.containsKey(hiveColumnIndex)) {
+                    domainFilters[i] = toFilter(domains.get(hiveColumnIndex));
+                    columnTypes[i] = columnHandle.getHiveType().getType(typeManager);
+                }
+            }
+        }
+
+        this.functionInputs = IntStream.range(0, columnMappings.size())
+                .boxed()
+                .collect(toImmutableMap(i -> columnMappings.get(i).getHiveColumnHandle().getHiveColumnIndex(), Function.identity()));
+
+        Map<VariableReferenceExpression, InputReferenceExpression> variableToInput = columnMappings.stream()
+                .map(HivePageSourceProvider.ColumnMapping::getHiveColumnHandle)
+                .collect(toImmutableMap(
+                        columnHandle -> new VariableReferenceExpression(columnHandle.getName(), columnHandle.getHiveType().getType(typeManager)),
+                        columnHandle -> new InputReferenceExpression(columnHandle.getHiveColumnIndex(), columnHandle.getHiveType().getType(typeManager))));
+
+        RowExpression optimizedRemainingPredicate = rowExpressionService.getExpressionOptimizer().optimize(remainingPredicate, OPTIMIZED, session);
+        if (TRUE_CONSTANT.equals(optimizedRemainingPredicate)) {
+            this.filterFunction = null;
+        }
+        else {
+            RowExpression expression = replaceExpression(optimizedRemainingPredicate, variableToInput);
+            this.filterFunction = new FilterFunction(session,
+                    rowExpressionService.getDeterminismEvaluator().isDeterministic(expression),
+                    rowExpressionService.getPredicateCompiler().compilePredicate(session.getSqlFunctionProperties(), expression).get());
+        }
+
+        this.outputBlockCount = requireNonNull(originalIndexes, "originalIndexes is null").size();
```

Comment:
@bhhari  nit: perhaps, verify that all of originalIndexes come from the start of the list of columns, e.g. 

```
verify(originalIndices.equals(columnMappings.stream().map(columnMapping -> columnMapping.getHiveColumnHandle().getHiveColumnIndex()).collect(toImmutableSet())))
```